File: src/host/js/rvm.simple.js
// #=# VM definitions
input = ");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y";

FALSE = [0,0,5]; TRUE = [0,0,5]; NIL = [0,0,5];

stack = 0;
push = (x) => ((stack = [x,stack,0]), true);
pop = () => { let x = stack[0]; stack = stack[1]; return x; };

pos = 0;
get_byte = () => input[pos++].charCodeAt(0);
get_code = () => { let x = get_byte()-35; return x<0 ? 57 : x; };
get_int = (n) => { let x = get_code(); n *= 46; return x<46 ? n+x : get_int(n+x-46); };

// #=# Primitives
fs = require("fs");

putchar = (c) => {
    let buffer = Buffer.alloc(1);
    buffer[0] = c;
    fs.writeSync(1, buffer, 0, 1);
    return c;
};

getchar_sync = () => {
    let buffer = Buffer.alloc(1);
    if (fs.readSync(0, buffer, 0, 1))
        return buffer[0];
    return -1;
};

getchar = () => {
    push(pos<input.length ? get_byte() : getchar_sync());
    return true; // indicate that no further waiting is necessary
};

prim1 = (f) => () => push(f(pop()));
prim2 = (f) => () => push(f(pop(),pop()));
prim3 = (f) => () => push(f(pop(),pop(),pop()));

bool2scm = (x) => x ? TRUE : FALSE;
scm2str = (r) => {
    let f = (c) => (c===NIL?"":String.fromCharCode(c[0])+f(c[1]));
    return f(r[0]);
};
is_rib = (x) => {
    if (x === undefined) console.log(stack);
    return x.length;
};

primitives = [
// @@(primitives (gen body)
  prim3((z, y, x) => [x, y, z]),                    //  @@(primitive (##rib a b c))@@
  prim1((x) => x),                                  //  @@(primitive (##id x))@@
  () => (pop(), true),                              //  @@(primitive (##arg1 x y))@@
  () => push([pop(),pop()][0]),                     //  @@(primitive (##arg2 x y))@@
  () => push([pop()[0],stack,1]),                   //  @@(primitive (##close rib))@@
  prim1((x) => bool2scm(is_rib(x))),             //  @@(primitive (##rib? rib) (use bool2scm))@@
  prim1((x) => x[0]),                               //  @@(primitive (##field0 rib))@@
  prim1((x) => x[1]),                               //  @@(primitive (##field1 rib))@@
  prim1((x) => x[2]),                               //  @@(primitive (##field2 rib))@@
  prim2((y, x) => (x[0]=y, true)),                          //  @@(primitive (##field0-set! rib))@@
  prim2((y, x) => (x[1]=y, true)),                          //  @@(primitive (##field1-set! rib))@@
  prim2((y, x) => (x[2]=y, true)),                          //  @@(primitive (##field2-set! rib))@@
  prim2((y, x) => bool2scm(x===y)),              //  @@(primitive (##eqv? x y) (use bool2scm))@@
  prim2((y, x) => bool2scm(x<y)),                //  @@(primitive (##< x y) (use bool2scm))@@
  prim2((y, x) => x+y),                             //  @@(primitive (##+ x y))@@
  prim2((y, x) => x-y),                             //  @@(primitive (##- x y))@@
  prim2((y, x) => x*y),                             //  @@(primitive (##* x y))@@
  prim2((y, x) => x/y|0),                           //  @@(primitive (##quotient x y))@@
  getchar,                                          //  @@(primitive (##getchar))@@
  prim1(putchar),                                   //  @@(primitive (##putchar c))@@
  () => pop() && halt(),//will crash with error on != 0 @@(primitive (##exit n))@@
// )@@
];
// --------------------------------------------------------------

// #=# Decode Symbol Table
symtbl = NIL;
n = get_int(0);
while (n-- > 0) symtbl=[[0,[NIL,0,3],2],symtbl,0]; // symbols with empty names

accum = NIL;
n = 0;
while (1) {
  c = get_byte();
  if (c == 44) { symtbl=[[0,[accum,n,3],2],symtbl,0]; accum = NIL; n = 0; continue; }
  if (c == 59) break;
  accum = [c,accum,0];
  n++;
}

symtbl = [[0,[accum,n,3],2],symtbl,0];

list_tail = (x,i) => i ? list_tail(x[1],i-1) : x;
symbol_ref = (n) => list_tail(symtbl,n)[0];

// #=# Decode instruction graph 
while (1) {
  x = get_code();
  n = x;
  d = 0;
  op = -1;
  while ((d=[20,30,0,10,11,4][++op])+2<n) n -= d+3;
  if (x>90)
    n = pop();
  else {
    if (!op) stack = [0,stack,0];
    n = n>=d ? (n==d ? get_int(0) : symbol_ref(get_int(n-d-1))) : op<3 ? symbol_ref(n) : n;
    if (4<op) {
      n = [[n,0,pop()],0,1];
      if (!stack) break;
      op=4;
    }
  }
  stack[0] = [op?op-1:0,n,stack[0]];
}

set_global = (x) => { symtbl[0][0] = x; symtbl = symtbl[1]; };

set_global([0,symtbl,1]); // primitive 0
set_global(FALSE);
set_global(TRUE);
set_global(NIL);

// #=# Execute RVM instructions

pc = n[0][2];

stack = [0,0,[5,0,0]]; // primordial continuation (executes halt instr.)


get_opnd = (o) => is_rib(o) ? o : list_tail(stack,o);
get_cont = () => { let s = stack; while (!s[2]) s = s[1]; return s; };


run = () => {
  while (1) {
    let o = pc[1];
    switch (pc[0]) {
    case 5: // halt
        return;
    case 0: // jump/call
        o = get_opnd(o)[0];
        while(1) {
            let c = o[0];

            if (is_rib(c)) {
                let nargs=pop();
                let c2 = [0,o,0];
                let s2 = c2;

                let nparams = c[0] >> 1;
                if (c[0] & 1 ? nparams > nargs : nparams != nargs){
                    console.log("*** Unexpected number of arguments nargs:", nargs, " nparams:", nparams, "variadics:", c[0]&1);
                    halt();
                }
                nargs-=nparams;
                if (c[0]&1) {
                    let rest=NIL;
                    while(nargs--)
                        rest=[pop(), rest, 0];
                    s2=[rest,s2,0];
                }
                while (nparams--) s2 = [pop(),s2,0];

                if (pc[2]===0) {
                    // jump
                    let k = get_cont();
                    c2[0] = k[0];
                    c2[2] = k[2];
                } else {
                    // call
                    c2[0] = stack;
                    c2[2] = pc[2];
                }
                stack = s2;
            } else {
                pop();
                o=primitives[c]();
                if (!o) return;
                if (is_rib(o)) continue;
                if (pc[2]===0) {
                    // jump
                    c = get_cont();
                    stack[1] = c[0];
                } else {
                    // call
                    c = pc;
                }
            }
            pc = c;
            break;
        }
        break;
    case 1: // set
        get_opnd(o)[0] = pop();
        break;
    case 2: // get
        push(get_opnd(o)[0]);
        break;
    case 3: // const
        push(o);
        break;
    case 4: // if
        if (pop() !== FALSE) { pc = pc[1]; continue; }
        break;
    }
      pc = pc[2];
  }
};

run(); 

File: src/host/py/rvm.simple.py
# #=# VM definitions
input=");'u?>vD?>vRD?>vRA?>vRA?>vR:?>vR=!(:lkm!':lkv6y" # RVM code that prints HELLO!

FALSE=[0,0,5]
TRUE=[0,0,5]
NIL=[0,0,5]

stack=0
def push(x):
 global stack
 stack=[x,stack,0]

def pop():
 global stack
 x=stack[0]
 stack=stack[1]
 return x

pos=-1
def get_byte():
 global pos
 pos+=1
 return ord(input[pos])

def get_code():
 x=get_byte()-35
 return 57 if x<0 else x

def get_int(n):
 x=get_code()
 n*=46
 return n+x if x<46 else get_int(n+x-46)


# #=# Primitives
import sys
stdo=sys.stdout

putchar=lambda c:[stdo.write(chr(c)),stdo.flush(),c][2]

def getchar():
 c=sys.stdin.read(1)
 push(ord(c) if len(c) else -1)

def prim(n, f):
    return lambda:push(f(*[pop() for _ in range(n)]))

prim1=lambda f: prim(1, f)
prim2=lambda f: prim(2, f)
prim3=lambda f: prim(3, f)

def f0s(y,x):x[0]=y;return y
def f1s(y,x):x[1]=y;return y
def f2s(y,x):x[2]=y;return y

def bool2scm(x):
    return TRUE if x else FALSE

def is_rib(x):
    return isinstance(x, list)

def scm2str(s):
 def chars2str(c):
  return (chr(c[0]) + chars2str(c[1])) if c is not NIL else "" 
 return chars2str(s[0])

primitives = [
 # @@(primitives (gen body) 
 prim(3,lambda z,y,x:[x,y,z]),                                            # @@(primitive (##rib a b c))@@           
 prim(1,lambda x:x),                                                      # @@(primitive (##id x))@@
 lambda:(pop(),None)[1],                                                  # @@(primitive (##arg1 x y))@@
 lambda:push([pop(),pop()][0]),                                           # @@(primitive (##arg2 x y))@@
 lambda:push([pop()[0],stack,1]),                                         # @@(primitive (##close rib))@@
 prim(1,lambda x:bool2scm(is_rib(x))),                                    # @@(primitive (##rib? rib))@@
 prim(1,lambda x:x[0]),                                                   # @@(primitive (##field0 rib))@@
 prim(1,lambda x:x[1]),                                                   # @@(primitive (##field1 rib))@@
 prim(1,lambda x:x[2]),                                                   # @@(primitive (##field2 rib))@@
 prim(2,f0s),                                                             # @@(primitive (##field0-set! rib x))@@
 prim(2,f1s),                                                             # @@(primitive (##field1-set! rib x))@@
 prim(2,f2s),                                                             # @@(primitive (##field2-set! rib x))@@
 prim(2,lambda y,x:bool2scm(x is y if is_rib(x) or is_rib(y) else x==y)), # @@(primitive (##eqv? x y))@@
 prim(2,lambda y,x:bool2scm(x<y)),                                        # @@(primitive (##< a b))@@
 prim(2,lambda y,x:x+y),                                                  # @@(primitive (##+ a b))@@
 prim(2,lambda y,x:x-y),                                                  # @@(primitive (##- a b))@@
 prim(2,lambda y,x:x*y),                                                  # @@(primitive (##* a b))@@
 prim(2,lambda y,x:int(x/y)),                                             # @@(primitive (##quotient a b))@@
 getchar,                                                                 # @@(primitive (##getchar))@@
 prim(1,putchar),                                                         # @@(primitive (##putchar c))@@
 prim(1,exit),                                                            # @@(primitive (##exit a))@@
# )@@ 
]

# #=# Decode Symbol Table
symtbl=NIL
n=get_int(0)
while n>0:
 n-=1
 symtbl=[[FALSE,[NIL,0,3],2],symtbl,0]

accum=NIL
n=0
while 1:
 c=get_byte()
 if c==44:
  symtbl=[[FALSE,[accum,n,3],2],symtbl,0]; accum=NIL; n=0
 else:
  if c==59: break
  accum=[c,accum,0]
  n+=1

list_tail=lambda lst,i:lst if i==0 else list_tail(lst[1],i-1)
symtbl=[[FALSE,[accum,n,3],2],symtbl,0]
symbol_ref=lambda n: list_tail(symtbl,n)[0]

# #=# Decode instruction graph 

LENGTH_ARRAY=[20, 30, 0, 10, 11, 4]
while True:
    token = get_code()
    arg = token
    op = 0
    range_val = LENGTH_ARRAY[op]
    while range_val + 2 < arg:
        arg -= range_val + 3
        op += 1
        range_val = LENGTH_ARRAY[op]
    if token > 90:
        arg = pop()
    else:
        if not op:
            stack = [0, stack, 0]
        if arg >= range_val:
            if arg == range_val:
                arg = get_int(0)
            else:
                arg = symbol_ref(get_int(arg - range_val - 1))
        else:
            if op < 3:
                arg = symbol_ref(arg)
        if op > 4:
            arg = [[arg, 0, pop()], 0, 1]
            if not stack:
                break
            op = 4
    if op != 0:
        stack[0] = [op - 1, arg, stack[0]]
    else:
        stack[0] = [0, arg, stack[0]]

def set_global(x):
    global symtbl
    symtbl[0][0] = x
    symtbl = symtbl[1]

set_global([0, symtbl, 1])
set_global(FALSE)
set_global(TRUE)
set_global(NIL)

# #=# Execute RVM instructions

# We can take the pc from the last closure created
pc = arg[0][2]
stack=[0,0,[5,0,0]] # primordial continuation (executes halt instr.)

get_opnd=lambda o:(o if is_rib(o) else list_tail(stack,o))

def get_cont():
 s=stack
 while not s[2]:s=s[1]
 return s

while 1:
 o=pc[1]
 i=pc[0]
 if i<1: # jump/call
  o=get_opnd(o)[0]
  while 1:
   c=o[0]
   if is_rib(c):
    nargs=pop(); 
    c2=[0,o,0]
    s2=c2
    nparams=c[0]>>1
    if nparams > nargs if c[0]&1 else nparams != nargs:
     print("*** Unexpected number of arguments nargs:", nargs, "nparams", nparams, "variadics:", c[0]&1);
     exit(1)
    nargs-=nparams
    if c[0]&1: 
     rest=NIL
     while nargs:
      rest=[pop(), rest, 0]
      nargs-=1
      
     s2=[rest,s2,0]
    while nparams:
     s2=[pop(),s2,0]
     nparams-=1

    if pc[2]: # call
     c2[0]=stack
     c2[2]=pc[2]
    else: # jump
     k=get_cont()
     c2[0]=k[0]
     c2[2]=k[2]
    stack=s2
   else:
    pop()
    o=primitives[c]()
    if is_rib(o): continue
    if pc[2]: # call
     c=pc
    else: # jump
     c=get_cont()
     stack[1]=c[0]
   pc=c
   break
 elif i<2: # set
  get_opnd(o)[0]=stack[0]; stack = stack[1]
 elif i<3: # get
  push(get_opnd(o)[0])
 elif i<4: # const
  push(o)
 elif i<5: # if
  if pop() is not FALSE:
   pc=pc[1]
   continue
 else: # halt
  break
 pc = pc[2]
